### PR TITLE
feat(server): Implement CLIENT PAUSE

### LINKED ISSUE #1609: Add support for CLIENT PAUSE command

### PR DESCRIPTION
**Pull request by royjacobson**
Close #1609 

The important part is in `server_state.cc`. Most of the changes in `server_family.cc` are the result of splitting `Client(...)` into sub-methods as it got too large.

### CODE REVIEW

File: src/server/server_family.h
Comment (romange) [inline src/server/server_family.h:185]: Weird case for these function names. Not sure it's by our styleguide book :)

File: src/server/main_service.cc
Comment (dranikpg) [inline src/server/main_service.cc:943]: This will produce inconsistent results with multi. MULTI/EXEC and many script will run till the end, yet some scripts that use DispatchCommand will block here 

I guess the intention is to let all multi run till the end - just check !under_script

Comment (dranikpg) [reply src/server/main_service.cc:943]: And it does not block DispatchManyCommands (which is probably not intended)

File: src/server/server_family.cc
Comment (dranikpg) [reply src/server/server_family.cc:1111]: Btw you might want to add a ToUpper before the Case()

File: src/server/main_service.cc
Comment (adiholden) [reply src/server/main_service.cc:922]: from redis doc: that interaction with replicas should continue normally 
can we check if the connection is from a replica and if so allow the command execution?

File: src/server/server_family.cc
Comment (royjacobson) [reply src/server/server_family.cc:1127]: Ok this is actually a bigger problem - I have to not block server shutdown somehow :/

**Comment by royjacobson** [comment]
This PR needs a bit more work. Like Adi has pointed out, this currently blocks the clients until the pause is cleared.

We can't easily put the unpausing into a new fiber and detach it: if we shutdown the server the new fiber has to be cancellable somehow and server shutdown has to wait for this cancellation, otherwise the unpause fiber could access invalid objects.

There's also another issue with replication that I haven't looked into yet.

File: tests/dragonfly/replication_test.py
Comment (dranikpg) [inline tests/dragonfly/replication_test.py:1720]: Let's avoid specific numbers for timing related tests. You can just start the seeder without a target and then call stop() and await the task as you already do at the end

**Review by dranikpg — DISMISSED** [review]

**Review by adiholden — APPROVED** [review]

File: src/server/server_family.cc
Comment (dranikpg) [inline src/server/server_family.cc:1344]: Please add a comment why it's impossible for DF to shutdown in less than 10ms leaving this fiber hanging. I guess because we wait on the pool to finish all tasks?

File: src/server/server_state.cc
Comment (dranikpg) [inline src/server/server_state.cc:151]: p*au*se 🙂

File: src/server/server_state.cc
Comment (dranikpg) [inline src/server/server_state.cc:143]: await (return !pause_dispatch_) 🙃

File: src/server/main_service.cc
Comment (dranikpg) [inline src/server/main_service.cc:1071]: 1. So if pause dispatch worked, this part would assume that all commands going through here are standalone? Why don't we just check instead of assuming?
2. Currently EXEC is always "write"

File: src/server/main_service.cc
Comment (dranikpg) [inline src/server/main_service.cc:1265]: This will stop commands from being collected, but they still will be executed

File: src/server/main_service.cc
Comment (dranikpg) [reply src/server/main_service.cc:1071]: 3. It's better to use Service::VerifyCommandState because squashed commands don't use DispatchCommand

File: src/server/server_family.cc
Comment (adiholden) [reply src/server/server_family.cc:1317]: Yes, SetPauseDispatch does not stop commands from dispatching, maybe the naming can be better here.
The idea here is to stop client from reading new data , wait untill all commands that we already read finish dispatching then set client pause and finally allow clients to read more data. 
This will ensure that after the last step of reading more data all clients read the new value of pause client, this is the reason for suspend dispatch. 
What is your suggestion? How will it help to check if there is mo client in the middle of a transaction? I want all transactions to stop before I change the pause state

File: src/server/main_service.cc
Comment (adiholden) [reply src/server/main_service.cc:1265]: Yes, this change was done by Roy, this is because of your comment about DispatchManyCommands that does not calls DispatchCommand, so this way the commands are not collected and therefore will be executed with DispatchCommand call bellow.

File: src/server/main_service.cc
Comment (adiholden) [reply src/server/main_service.cc:1071]: 1. by standalone do you mean not under script or multi transaction? Well no I do not assume this.. what I assume is that I am not in the middle of a transaction, i.e I did not execute few commands in multi and then client pause state was set and then I wait till pause is finished and will continue executing some other commands in the transaction.
2. why are you saying this? if exec has non wirte commands inside it is not write.. I just verified that with cli.
3. we do not enable squashed commands when pause is set, see my comment on what you wrote on collecting commands bellow.

**Comment by dranikpg** [comment]
> by standalone do you mean not under script or multi transaction? Well no I do not assume this.. what I assume is that I am not in the middle of a transaction, i.e I did not execute few commands in multi and then client pause state was set and then I wait till pause is finished and will continue executing some other commands in the transaction.

Yes, but a multi transaction is always a script or a multi/exec command. I'm trying to understand why we need this preliminary waiting state at all
 
If this snippet won't block already running multi transactions, but it will prevent any _new_ ones from running
```
if (!exec && !script) {
  AwaitPaused()
}
```

**Comment by dranikpg** [comment]
Ohh, okay, you need to guarantee this by replying to the client... forgot about this 😅 Now I understand why we need to wait

I can suggest #2156, so we don't have one more solution for this. Once I finished I can add this here. If you decide you can also go ahead with your approach

**Comment by adiholden** [comment]
> Ohh, okay, you need to guarantee this by replying to the client... forgot about this 😅 Now I understand why we need to wait
> 
> I can suggest #2156, so we don't have one more solution for this. Once I finished I can add this here. If you decide you can also go ahead with your approach

But still this PR #2156 will not replace the logic to stop dispatching commands inorder to set the value of client puase for read or write, it will only replace my call for AwaitDispatches in my PR

**Comment by dranikpg** [comment]
I see, the real issue is that when a command is blocked on `AwaitPauseState` it is in the dispatching state. That's actually not easy to manage because we can't await dispatches for that, we have to block the flow of commands... Then I'm not gonna interfere anymore 😅

**Review by dranikpg — DISMISSED** [review]

File: src/server/CMakeLists.txt
Comment (romange) [inline src/server/CMakeLists.txt:75]: why this change?

File: src/server/main_service.cc
Comment (romange) [inline src/server/main_service.cc:1075]: you can track it when you add a command to the body.

Comment (romange) [reply src/server/main_service.cc:1075]: and store it in exec_info

File: src/server/server_state.h
Comment (romange) [inline src/server/server_state.h:236]: there are three commands with similar names, I suggest adding comments on how they differ

File: src/server/CMakeLists.txt
Comment (adiholden) [reply src/server/CMakeLists.txt:75]: looks like issue with rebasing the branch, as at the time this PR was created there was no server_family_test
will fix this

File: src/server/server_family.cc
Comment (romange) [inline src/server/server_family.cc:1325]: puased -> paused

Comment (romange) [reply src/server/server_family.cc:1325]: why we should not expire/evict keys?

File: src/server/server_family.cc
Comment (romange) [inline src/server/server_family.cc:1329]: I do not understand what's the purpose of this fiber

Comment (adiholden) [reply src/server/server_family.cc:1329]: This fiber stop the client pause after timeout is reached.
We need to use a fiber as we need to return OK to client ASAP , so the CLIENT PAUSE command execution is not paused by itself.

**Comment by adiholden** [comment]
@romange do I get an merge approval?

**Review by romange — APPROVED** [review]

**Pull request merged**
